### PR TITLE
SR-1005: [Foundation] Replace posix_spawn by NSTask

### PR DESCRIPTION
**Pull request by Bouke**
Currently this cannot be merged as builds fail on Linux;

* [x] Builds on Linux need to include Foundation ([PR][1])
* [x] SwiftFoundation doesn't implement output redirection (`NSTask.standardOutput`), so no output can be captured ([PR][2]).
* [x] SwiftFoundation doesn't implement output streaming (`NSFileHandle.availableData`) ([PR][2]).
* [x] SwiftFoundation's NSTask fails to run some commands with custom environment variables ([PR][3]).
* [x] SwiftFoundation's NSTask `argv[0]` and stderr apple/swift-corelibs-foundation#353

[1]: https://github.com/apple/swift/pull/2406
[2]: https://github.com/apple/swift-corelibs-foundation/pull/348
[3]: https://github.com/apple/swift-corelibs-foundation/pull/349

### CODE REVIEW

**Comment by mxcl** [comment]
@swift-ci Please test

**Comment by Bouke** [comment]
The failing OS X build doesn't appear to be related. Regarding the failing Linux build, this PR is dependent on a change in Foundation ([PR][1]).

[1]: https://github.com/apple/swift-corelibs-foundation/pull/348

**Comment by Bouke** [comment]
When testing this on Linux, most build-related tests fail. What I've found so far is that `Process.arguments` contained the absolute path of the test suite. However after my patch, it somehow only contains the filename, not the path. As `Process.arguments` is used to infer the build directory, this fails as it is assumed to be in the CWD. For example see: [`swiftBuildPath`][1].

BEFORE:

    Process.arguments: 
    ["<builddir>/Ninja-ReleaseAssert/swiftpm-linux-x86_64/debug/SwiftPMTests.xctest"]

    swiftBuildPath: 
    "<builddir>/Ninja-ReleaseAssert/swiftpm-linux-x86_64/debug/swift-build"

AFTER:

    Process.arguments: 
    ["SwiftPMTests.xctest"]

    swiftBuildPath: 
    "<sourcedir>/swiftpm/swift-build"

[1]: https://github.com/apple/swift-package-manager/blob/22351877c847a20a0dd809590a91f4f8d30bc8ce/Tests/Functional/Utilities.swift#L109

**Comment by Bouke** [comment]
I've rebased the PR and tested against OSX and Ubuntu.

File: Sources/Utility/system.swift [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    let task = NSTask()
+
+    var arguments = arguments
+    task.launchPath = which(arguments.removeFirst())
+    task.arguments = arguments
+
+    var environment = environment
+    #if Xcode
+        let keys = ["SWIFT_EXEC", "HOME", "PATH"]
```
Comment (ddunbar) [inline Sources/Utility/system.swift]: This lost the change to add: `"TOOLCHAINS", ` here that I made in the past couple of days.

File: Sources/Utility/which.swift [hunk 2]
```diff
@@ -10,10 +10,15 @@
...
 
+private var memoized = [String: String]()
+
 public func which(_ arg0: String) -> String {
     if arg0.isAbsolute {
         return arg0
+    } else if let fullpath = memoized[arg0] {
+        return fullpath
     } else if let fullpath = try? Utility.popen(["/bin/sh", "-c", "which \(arg0)"]) {
+        memoized[arg0] = fullpath.chomp()
```
Comment (ddunbar) [inline Sources/Utility/which.swift]: We should probably just implement the lookup directly rather than actually shell out for this relatively trivial task.

File: Sources/ManifestSerializer/Manifest+parse.swift [hunk 3]
```diff
@@ -66,17 +67,17 @@ private func parse(path manifestPath: String, swiftc: String, libdir: String) th
...
+    //allow child process to access parent's file descriptors. So instruct
+    //swiftc to dump the package information on stdout.
+    #if os(OSX) || os(iOS)
+        let handle = NSFileHandle.standardOutput()
+    #else
+        let handle = NSFileHandle.fileHandleWithStandardOutput()
+    #endif
+
+    cmd += ["-fileno",  String(handle.fileDescriptor)]
+    let toml = try Utility.popen(cmd)
```
Comment (ddunbar) [inline Sources/ManifestSerializer/Manifest+parse.swift]: This is rather unfortunate, it was actually a benefit that we could do this through a special file descriptor which also meant logging output could go on stdout and not break anything.

File: Sources/Utility/which.swift (hunk 2 above)
Comment (Bouke) [inline Sources/Utility/which.swift]: I've added this in a new commit. It's a nice way as well to improve the test suite performance;

| Context                     | Linux | OS X  |
| --------------------------- | ----- | ----- |
| Baseline (master@827b703)   | 105 s | 141 s |
| Shell out to which          | 104 s | 142 s |
| lookup PATH                 | 110 s | 144 s |
| lookup PATH + memo          | 104 s | 140 s |

File: Sources/Utility/system.swift (hunk 1 above)
Comment (Bouke) [inline Sources/Utility/system.swift]: Yes I missed that, I've restored those changes.

File: Sources/ManifestSerializer/Manifest+parse.swift (hunk 3 above)
Comment (Bouke) [inline Sources/ManifestSerializer/Manifest+parse.swift]: The problem here is that NSTask (Cocoa) closes all file descriptors on the child process. A different way would be to pass the path of a temporary file to which the toml is written. Is that fine as well?

File: Sources/Utility/which.swift [hunk 4]
```diff
@@ -9,19 +9,30 @@
 */
 
 import Foundation
+import func POSIX.getenv
+import func POSIX.getcwd
 
-private var memoized = [String: String]()
+private let PATH = [POSIX.getcwd()] + (getenv("PATH") ?? "").components(separatedBy: ":")
```
Comment (ddunbar) [inline Sources/Utility/which.swift]: We shouldn't default to looking in the `cwd`, that can be a security issue.

File: Sources/Utility/which.swift [hunk 5]
```diff
@@ -9,19 +9,30 @@
...
-        return fullpath.chomp()
-    } else {
-        return arg0
     }
+    if let path = memo[arg0] {
+        return path
+    }
+    for prefix in PATH {
+        let path = Path.join([prefix, arg0])
+        if path.isFile {
```
Comment (ddunbar) [inline Sources/Utility/which.swift]: This should also check if it is executable.

File: Sources/Utility/which.swift [hunk 6]
```diff
@@ -9,19 +9,30 @@
...
-    } else if let fullpath = try? Utility.popen(["/bin/sh", "-c", "which \(arg0)"]) {
-        memoized[arg0] = fullpath.chomp()
-        return fullpath.chomp()
-    } else {
-        return arg0
     }
+    if let path = memo[arg0] {
+        return path
+    }
+    for prefix in PATH {
```
Comment (ddunbar) [inline Sources/Utility/which.swift]: We should follow the standard shell behavior in this function, where if `arg0` contains any `/` then we use it directly, regardless of its existence.

File: Sources/Utility/which.swift [hunk 7]
```diff
@@ -9,19 +9,30 @@
...
+        let path = Path.join([prefix, arg0])
+        if path.isFile {
+            memo[arg0] = path
+            return path
+        }
+    }
+    throw Error.UnknownCommand(arg0: arg0)
 }
-
```
Comment (ddunbar) [inline Sources/Utility/which.swift]: This function is generally useful infrastructure, can we land this in a separate PR?

File: Sources/ManifestSerializer/Manifest+parse.swift (hunk 3 above)
Comment (ddunbar) [inline Sources/ManifestSerializer/Manifest+parse.swift]: It just doesn't seem worth it to me. I think this is a case where it makes the most sense to use posix_spawn directly rather than work harder just to use `NSTask`.

**Comment by ddunbar** [comment]
I'm closing this out... per my email to swift-build-dev I'm not longer convinced we should try to move all of our infrastructure onto the Foundation APIs. I think instead we should do it on a case by case basis, and in this case (posix_spawn) my feeling is that our needs match using the POSIX API directly, and I am fine with us needing to maintain the infrastructure for that (at least until there is a standard POSIX module for Swift itself).

I think some of the cleanups on this branch make sense, but it didn't look like I could easily cherry pick them out.
